### PR TITLE
Klax 0.2.0

### PR DESCRIPTION
**Pull request by Drenderer**
# Training 
- [x] Split the training loop from `klax.fit` into its major components: 
  - Loss and loss-gradient computation
  - Optimisation: weight update computation
  - Callbacks
  - One loop that glues these components together.
- [x] Turn the current `CallbackArgs` class into a `TrainingState` class (ideally as PyTree via `eqx.Module`). This can then act as the interaction object, tying together the components of the training loop.
- [x] Fix the `batcher` vs `BatchGenerator` vs `batch` naming issue.
- [x] Generalize the bach_axes argument of a `Loss` to an arbitrary, user-defined `auxiliary_state` or `training_context` pytree to the `TrainingState`.
- [ ] (Optional) _Loss aux_: Generalize `klax.Loss` to return a tuple `(loss_value, loss_aux)`. The auxiliary output `loss_aux` is an arbitrary PyTree that is not used in the gradient computation or model update, but it is passed to any callbacks. It is useful to compute metrics without recomputation of the model activations - for example, to directly log the training loss on the current batch, without recomputing it, by returning it with the loss from the loss function. The user could still define "traditional" loss functions with a single scalar output if `klax.Loss` canonicalizes them. 
The only issue that would result from this generalization is that Loss functions could not easily be used as functions for the `BatchMetric`.

# Models/Architectures
- [x] Implement [weight initialisation for ICNNs](https://proceedings.neurips.cc/paper_files/paper/2023/file/9062b7d6e522dadf4f7d85d49b60d81e-Paper-Conference.pdf)
- [x] Implement [PICNN](https://arxiv.org/pdf/1609.07152)
- [x] Refactor the FICNN implementation by implementing a FICNNLayer to be in line with the PICNN implementation.

# Docs
### Structure
- [ ] Update the fundamental structure of the docs into user-facing APIs and advanced APIs.
### Explanations 
- [ ] Description of `klax.fit` and its capabilities for customisation. This can be tied together with the following examples/tutorials
  - [ ] Example/ Tutorial on writing custom loss functions.
  - [ ] Example/ Tutorial on using constraints.
  - [x] Example/ Tutorial on writing custom callbacks.

### CODE REVIEW

**Review by jaosch — COMMENTED** [review]
Ready for merge.

**Review by jaosch — APPROVED** [review]
